android-interop-testing/app/src/main/java/io/grpc/android/integrationtest/InteropTester.java

```diff
@@ -0,0 +1,346 @@
+    StreamObserver<Messages.StreamingInputCallRequest> requestObserver =
+        asyncStub.streamingInputCall(responseObserver);
+    for (Messages.StreamingInputCallRequest request : requests) {
+      requestObserver.onValue(request);
+    }
+    requestObserver.onCompleted();
+    assertEquals(goldenResponse, responseObserver.firstValue().get());
+  }
+
+  public void pingPong() throws Exception {
+    final Messages.StreamingOutputCallRequest[] requests =
+        new Messages.StreamingOutputCallRequest[4];
+    for (int i = 0; i < 4; i++) {
+      requests[i] = new Messages.StreamingOutputCallRequest();
+      requests[i].responseParameters = new Messages.ResponseParameters[1];
+      requests[i].responseParameters[0] = new Messages.ResponseParameters();
+      requests[i].payload = new Messages.Payload();
+    }
+    requests[0].responseParameters[0].size = 31415;
+    requests[0].payload.body = new byte[27182];
+    requests[1].responseParameters[0].size = 9;
+    requests[1].payload.body = new byte[8];
+    requests[2].responseParameters[0].size = 2653;
+    requests[2].payload.body = new byte[1828];
+    requests[3].responseParameters[0].size = 58979;
+    requests[3].payload.body = new byte[45904];
+
+
+    final Messages.StreamingOutputCallResponse[] goldenResponses =
+        new Messages.StreamingOutputCallResponse[4];
+    for (int i = 0; i < 4; i++) {
+      goldenResponses[i] = new Messages.StreamingOutputCallResponse();
+      goldenResponses[i].payload = new Messages.Payload();
+      goldenResponses[i].payload.type = Messages.COMPRESSABLE;
+    }
+    goldenResponses[0].payload.body = new byte[31415];
+    goldenResponses[1].payload.body = new byte[9];
+    goldenResponses[2].payload.body = new byte[2653];
+    goldenResponses[3].payload.body = new byte[58979];
+
+    final LinkedBlockingQueue<Object> responses = new LinkedBlockingQueue<>();
+    final Object magicTailResponse = new Object();
+    @SuppressWarnings("unchecked")
+    StreamObserver<Messages.StreamingOutputCallResponse> responseObserver =
+        new StreamObserver<Messages.StreamingOutputCallResponse>() {
+
+          @Override
+          public void onValue(Messages.StreamingOutputCallResponse value) {
+            responses.add(value);
+          }
+
+          @Override
+          public void onError(Throwable t) {
+            Log.e(TesterActivity.LOG_TAG, "Encounter an error", t);
+            responses.add(t);
+          }
+
+          @Override
+          public void onCompleted() {
+            responses.add(magicTailResponse);
+          }
+        };
+    StreamObserver<Messages.StreamingOutputCallRequest> requestObserver
+        = asyncStub.fullDuplexCall(responseObserver);
+    for (int i = 0; i < requests.length; i++) {
+      requestObserver.onValue(requests[i]);
+      Object response = responses.poll(5, TimeUnit.SECONDS);
+      Assert.assertFalse(response.toString(), response instanceof Throwable);
```

Comment:
Done.

And I measured the time (roughly time from I clicked enter on my keyboard to the result was outputted) on a real device: without toString, it took ~3 seconds for ping_pong test, while ~9 seconds was used with toString. I think it will get worse on an emulator.